### PR TITLE
Implement experience

### LINKED ISSUE #457: Implement XP.

### PR DESCRIPTION
**Pull request by JustTalDevelops**
Closes #457.

### CODE REVIEW

**Comment by JustTalDevelops** [comment]
This is missing orb pickup cooldowns (100ms) but aside from that this should be ready to review.

**Comment by JustTalDevelops** [comment]
This should be done completely, just waiting on a review.

File: server/entity/experience.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+	"math"
+	"sync"
+)
+
+// ExperienceManager manages experience and levels for entities, and provides functions to add, remove, and calculate
+// experience needed for upcoming levels.
+type ExperienceManager struct {
+	mu         sync.RWMutex
+	total      int
+	experience int
```
Comment (Sandertv) [inline server/entity/experience.go:14]: Do you reckon we could actually only have the `total` field here and calculate the per-level `experience` on the fly where necessary?

Comment (JustTalDevelops) [reply server/entity/experience.go:14]: Pretty sure that total is for lifetime experience

File: server/player/player.go [hunk 2]
```diff
@@ -2374,6 +2470,9 @@ func (p *Player) Data() Data {
 		Health:          p.Health(),
 		MaxHealth:       p.MaxHealth(),
 		Hunger:          p.hunger.foodLevel,
+		XPLevel:         p.ExperienceLevel(),
+		XPPercentage:    p.ExperienceProgress(),
+		XPTotal:         p.TotalLifetimeExperience(),
```
Comment (Sandertv) [inline server/player/player.go:2475]: Would replace these with a single field and store `p.Experience()`.

File: server/entity/experience.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	"fmt"
+	"math"
+	"sync"
+)
+
+// ExperienceManager manages experience and levels for entities, and provides functions to add, remove, and calculate
+// experience needed for upcoming levels.
+type ExperienceManager struct {
+	mu         sync.RWMutex
+	total      int
```
Comment (Sandertv) [inline server/entity/experience.go:13]: Would remove the lifetime stuff as discussed.

**Review by Sandertv — APPROVED** [review]
This looks good now! Thanks!

**Pull request merged**